src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -2465,38 +2465,55 @@ String ReplicatedMergeTreeMergePredicate::getCoveringVirtualPart(const String &
 ReplicatedMergeTreeQueue::SubscriberHandler
 ReplicatedMergeTreeQueue::addSubscriber(ReplicatedMergeTreeQueue::SubscriberCallBack && callback)
 {
-    std::lock_guard lock(state_mutex);
-    std::unordered_set<String> result;
-    result.reserve(queue.size());
-    for (const auto & entry : queue)
-        result.insert(entry->znode_name);
-
+    std::lock_guard<std::mutex> lock(state_mutex);
     std::lock_guard lock_subscribers(subscribers_mutex);
 
     auto it = subscribers.emplace(subscribers.end(), std::move(callback));
 
-    /// Notify queue size & log entry ids to avoid waiting for removed entries
-    (*it)(result.size(), result, std::nullopt);
+    /// Atomically notify about current size
+    (*it)(queue.size(), nullptr);
 
     return SubscriberHandler(it, *this);
 }
 
-ReplicatedMergeTreeQueue::SubscriberHandler::~SubscriberHandler()
+std::unordered_set<String> ReplicatedMergeTreeQueue::getEntryNamesSet(bool lightweight_entries_only)
 {
-    std::lock_guard lock(queue.subscribers_mutex);
-    queue.subscribers.erase(it);
+    std::lock_guard lock(state_mutex);
+    std::unordered_set<String> result;
+    result.reserve(queue.size());
+    for (const auto & entry : queue)
+    {
+        bool is_lightweight = entry->type == LogEntry::GET_PART || entry->type == LogEntry::ATTACH_PART
+            || entry->type == LogEntry::DROP_RANGE || entry->type == LogEntry::REPLACE_RANGE || entry->type == LogEntry::DROP_PART;
```

Comment:
I would prefer creating `static constexpr std::array` and doing `std::any_of` because this is huge.
But I'm fine either way.